docs/standard/microservices-architecture/multi-container-microservice-net-applications/implementing-api-gateways-with-ocelot.md

```diff
@@ -0,0 +1,579 @@
+
+```csharp
+[HttpGet]
+[Route("items/{id:int}")]
+[ProducesResponseType((int)HttpStatusCode.NotFound)]
+[ProducesResponseType(typeof(CatalogItem),(int)HttpStatusCode.OK)]
+public async Task<IActionResult> GetItemById(int id)
+{
+    if (id <= 0)
+    {
+        return BadRequest();
+    }
+    var item = await _catalogContext.CatalogItems.
+                                          SingleOrDefaultAsync(ci => ci.Id == id);
+    //…
+
+    if (item != null)
+    {
+        return Ok(item);
+    }
+    return NotFound();
+}
+```
+The Http request will end up running that kind of C# code accessing the microservice database plus any additional action.
+
+In regards the microservice URL, when the containers are deployed in your local development PC (local Docker host), each microservice’s container has always an internal port, usually port 80, specified in its dockerfile, as in the following dockerfile.
+
+```csharp
+FROM microsoft/aspnetcore:2.0.5 AS base
+WORKDIR /app
+EXPOSE 80
+```
+The port 80 shown in the code is internal within the Docker host, so it cannot be reached by the client apps. 
+The client apps can access only to the external ports (if any) published when deploying with docker-compose.
+
+Those external ports should not be published when deploying into a production environment. This is precisely why you want to use the API Gateway, to avoid the direct communication between the client apps and the microservices.
+
+However, when developing, you want to access the microservice/container directly and run it through Swagger. That’s why in eShopOnContainers the external ports are still specified even when they won’t be used by the API Gateway or the client apps.
+
+Here’s an example of the docker-compose.override.yml file for the Catalog microservice.
+
+```csharp
+catalog.api:
+  environment:
+    - ASPNETCORE_ENVIRONMENT=Development
+    - ASPNETCORE_URLS=http://0.0.0.0:80
+    - ConnectionString=YOUR_VALUE
+    - ... Other Environment Variables
+  ports:
+    - "5101:80"   # Important: In a production environment you should remove the external port (5101) kept here for microservice debugging purposes. 
+                  # The API Gateway redirects and access through the internal port (80).
+```
+
+You can see how in the docker-compose.override.yml configuration the internal port for the Catalog container is port 80, but the port for external access is 5101. But this port shouldn’t be used by the application when using an API Gateway, only to debug run and test just the Catalog microservice.
+
+Normally, you won’t be deploying with docker-compose into a production environment because the right production deployment environment for microservices is an orchestrator like Kubernetes or Service Fabric and when deploying to those environments you will use different configuration files where you won’t publish directly any external port for the microservices but you will always use the reverse proxy from the API Gateway.
```

Comment:
Break the sentence in the middle. 

and when deploying to those environments you -> When deploying to those environments, you

add comma before but

you will always use -> you'll always use

Reply:
👍